Incremental Delivery spec draft

robrichard opened the pull request:
Spec draft of `@defer` based on the response format in https://github.com/graphql/defer-stream-wg/discussions/69, with no duplication of fields. `@stream` will be added soon.

Please keep discussion limited to specifics on how the spec text is written. There is a dedicated discussion forum for questions, feedback, and comments on defer & stream: https://github.com/graphql/defer-stream-wg/discussions

owen-zora commented:
Hi @robrichard , do you have any expected timeline for this feature? We are really excited to have this feature for our product!

yaacovCR commented on spec/Section 6 -- Execution.md:
```diff
@@ -513,14 +905,24 @@ CollectFields(objectType, selectionSet, variableValues, visitedFragments):
+      - Let {deferDirective} be that directive.
+      - If this execution is for a subscription operation, raise a _field
+        error_.
+    - If {deferDirective} is not defined:
```
I think this line can be removed. See https://github.com/graphql/graphql-js/pull/4320

robrichard committed:
Spec edits for incremental delivery, Section 3 only (#1132)

* Spec edits for incremental delivery, Section 3

* re-order arguments

* Apply suggestions from code review

Co-authored-by: Benjie <benjie@jemjie.com>

* Address PR feedback

* Address PR feedback

* Update spec/Section 3 -- Type System.md

Co-authored-by: Benoit 'BoD' Lubek <BoD@JRAF.org>

* Update response terminology

* Use new execution result terminology

* Update order of arguments on @stream

* field error -> execution error

* update argument order in examples

* _subsequent execution result_ > _execution update result_

* Update note to normative section

* fix title and link

* review feedback

---------

Co-authored-by: Benjie <benjie@jemjie.com>
Co-authored-by: Benoit 'BoD' Lubek <BoD@JRAF.org>

robrichard committed:
Spec edits for incremental delivery, Section 3 only (#1132)

* Spec edits for incremental delivery, Section 3

* re-order arguments

* Apply suggestions from code review

Co-authored-by: Benjie <benjie@jemjie.com>

* Address PR feedback

* Address PR feedback

* Update spec/Section 3 -- Type System.md

Co-authored-by: Benoit 'BoD' Lubek <BoD@JRAF.org>

* Update response terminology

* Use new execution result terminology

* Update order of arguments on @stream

* field error -> execution error

* update argument order in examples

* _subsequent execution result_ > _execution update result_

* Update note to normative section

* fix title and link

* review feedback

---------

Co-authored-by: Benjie <benjie@jemjie.com>
Co-authored-by: Benoit 'BoD' Lubek <BoD@JRAF.org>

michaelstaib commented on spec/Section 6 -- Execution.md:
```diff
@@ -325,39 +315,405 @@ Unsubscribe(responseStream):
+
+- If {serial} is not provided, initialize it to {false}.
+- Let {groupedFieldSet} and {newDeferUsages} be the result of
+  {CollectFields(objectType, selectionSet, variableValues)}.
```
The new `CollectFields` requires a `deferUsage` argument ... this should be `null` for the initial call to `CollectFields`.

```suggestion
  {CollectFields(objectType, selectionSet, null, variableValues)}.
```

Also in this case its not clear while reading what null actually is for so maybe it should be declared first?

yaacovCR committed:
execution edits for defer and stream

robrichard committed:
Spec edits for incremental delivery, Validation

robrichard committed:
Spec edits for incremental delivery, Response Section and Examples Appendix (#1203)

* Spec edits for incremental delivery, Examples

* Spec edits for incremental delivery, Section 7 only

* fix typos

* Update to address feedback in exampled appendix

* Update terminology

Initial Execution Result > Initial Incremental Stream Result
Execution Update Result > Incremental Stream Update Result

* PR feedback

* allow hasNext: false on initial result

* pending is must on initial stream result

* value => payload when describing what is yielded by a stream

* Update description of _initial incremental stream result_ & _incremental stream update result_

* fix incorrect term

* PR Feedback

* Pending Result -> Incremental Pending Notice

* Completed Result -> Incremental Completion Notice

robrichard committed:
Spec edits for incremental delivery, Response Section and Examples Appendix (#1203)

* Spec edits for incremental delivery, Examples

* Spec edits for incremental delivery, Section 7 only

* fix typos

* Update to address feedback in exampled appendix

* Update terminology

Initial Execution Result > Initial Incremental Stream Result
Execution Update Result > Incremental Stream Update Result

* PR feedback

* allow hasNext: false on initial result

* pending is must on initial stream result

* value => payload when describing what is yielded by a stream

* Update description of _initial incremental stream result_ & _incremental stream update result_

* fix incorrect term

* PR Feedback

* Pending Result -> Incremental Pending Notice

* Completed Result -> Incremental Completion Notice

robrichard committed:
Spec edits for incremental delivery, Validation (#1223)

* Spec edits for incremental delivery, Validation

* Update spec/Section 5 -- Validation.md

Co-authored-by: Yaacov Rydzinski  <yaacovCR@gmail.com>

* Update labels are unique rule to account for null value of label

* PR feedback

* Rewrite Valid Root Field rule

* review fixes

* rewrite Valid Operations rule

* review fixes

---------

Co-authored-by: Yaacov Rydzinski <yaacovCR@gmail.com>

robrichard committed:
Fixes for incremenal delivery integration branch (#1232)

robrichard committed:
Spec edits for incremental delivery, Execution: CollectFields

Co-authored-by: Yaacov Rydzinski <yaacovCR@gmail.com>

robrichard committed:
execution edits for defer and stream

Co-authored-by: Yaacov Rydzinski <yaacovCR@gmail.com>
